src/iface/interface/mod.rs

```diff
@@ -474,25 +474,16 @@ impl Interface {
         #[cfg(any(feature = "medium-ethernet", feature = "medium-ieee802154"))]
         let hardware_addr = match caps.medium {
             #[cfg(feature = "medium-ethernet")]
-            Medium::Ethernet => Some(
-                config
-                    .hardware_addr
-                    .expect("hardware_addr required option was not set"),
-            ),
-            #[cfg(feature = "medium-ip")]
-            Medium::Ip => {
-                assert!(
-                    config.hardware_addr.is_none(),
-                    "hardware_addr is set, but device medium is IP"
-                );
-                None
-            }
+            Medium::Ethernet => config
+                .hardware_addr
+                .expect("hardware_addr required option was not set"),
             #[cfg(feature = "medium-ieee802154")]
-            Medium::Ieee802154 => Some(
-                config
-                    .hardware_addr
-                    .expect("hardware_addr required option was not set"),
-            ),
+            Medium::Ieee802154 => config
+                .hardware_addr
+                .expect("hardware_addr required option was not set"),
+
+            #[cfg(feature = "medium-ip")]
+            Medium::Ip => unreachable!(),
```

Comment:
This is not unreachable. 

If the user enables both `medium-ethernet` and ``medium-ip` features, we want the user to be able to create both Ethernet and IP medium interfaces, choosing at runtime. (an example use case is if you have a device with both Ethernet and mobile connectivity).

This is the reason the `hardware_addr` was an `Option`. 

I agree the current solution is weird though. Perhaps it makes sense to make `HardwareAddress` be an Enum with `Ethernet, `Ieee802154` and `Ip` cases, where `Ip` is empty? (And move it from `wire` to `iface`, because that makes it very iface-specific)

Reply:
Indeed, sounds like the best solution at the moment.